precompiles/bank/method_deposit.go

```diff
@@ -0,0 +1,197 @@
+package bank
+
+import (
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+
+	ptypes "github.com/zeta-chain/node/precompiles/types"
+	"github.com/zeta-chain/node/x/fungible/types"
+)
+
+// deposit is used to deposit ZRC20 into the bank contract, and receive the same amount of cosmos coins in exchange.
+// The denomination of the cosmos coin will be "zrc20/ZRC20Address", as an example depossiting an arbitrary ZRC20 token with
+// address 0x12345 will mint cosmos coins with the denomination "zrc20/0x12345".
+// The caller cosmos address will be calculated from the EVM caller address. by executing toAddr := sdk.AccAddress(addr.Bytes()).
+// This function can be think of a permissionless way of minting cosmos coins.
+// This is how deposit works:
+// - The caller has to allow the bank contract to spend a certain amount ZRC20 token coins on its behalf. This is mandatory.
+// - Then, the caller calls deposit(ZRC20 address, amount), to deposit the amount and receive cosmos coins.
+// - The bank will check there's enough balance, the caller is not a blocked address, and the token is a not paused ZRC20.
+// - Then the cosmos coins "zrc20/0x12345" will be minted and sent to the caller's cosmos address.
+// Call this function using solidity with the following signature:
+// - From IBank.sol: function deposit(address zrc20, uint256 amount) external returns (bool success);
+func (c *Contract) deposit(
+	ctx sdk.Context,
+	evm *vm.EVM,
+	contract *vm.Contract,
+	method *abi.Method,
+	args []interface{},
+) (result []byte, err error) {
+	// This function is developed using the Check - Effects - Interactions pattern:
+	// 1. Check everything is correct.
+	if len(args) != 2 {
+		return nil, &(ptypes.ErrInvalidNumberOfArgs{
+			Got:    len(args),
+			Expect: 2,
+		})
+	}
+
+	// Unpack parameters for function deposit.
+	// function deposit(address zrc20, uint256 amount) external returns (bool success);
+	zrc20Addr, amount, err := unpackDepositArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the correct caller address.
+	caller, err := getEVMCallerAddress(evm, contract)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the cosmos address of the caller.
+	toAddr, err := getCosmosAddress(c.bankKeeper, caller)
+	if err != nil {
+		return nil, err
+	}
+
+	// Safety check: token has to be a valid whitelisted ZRC20 and not be paused.
+	t, found := c.fungibleKeeper.GetForeignCoins(ctx, zrc20Addr.String())
+	if !found || t.Paused {
+		return nil, &ptypes.ErrInvalidToken{
+			Got:    zrc20Addr.String(),
+			Reason: "token is not a whitelisted ZRC20 or it's paused",
```

Comment:
Can we rather give a separate error for not found and paused?

Reply:
Modified in af858abd8aaa8c7087e6ec75abe7299e34279220